js/util/pipeline-fixed.js

```diff
@@ -0,0 +1,335 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ * A copy of the GNU Lesser General Public License is in the file COPYING.
+ */
+
+/** @ignore */
+var Interest = require('../interest.js').Interest; /** @ignore */
+var Blob = require('./blob.js').Blob; /** @ignore */
+var KeyChain = require('../security/key-chain.js').KeyChain; /** @ignore */
+var NdnCommon = require('./ndn-common.js').NdnCommon;
+var DataFetcher = require('./data-fetcher.js').DataFetcher;
+var LOG = require('../log.js').Log.LOG;
+
+
+/**
+ * Retrieve the segments of solicited data by keeping a fixed-size window of N
+ * of on fly Interests at any given time.
+ *
+ * To handle timeout and nack we use DataFetcher class which upon facing
+ * timeout or nack will try to resolve the corresponded segment by retransmitting
+ * the Interest a few times.
+ *
+ * PipelineFixed assumes that the data is named /<prefix>/<version>/<segment>,
+ * where:
+ * - <prefix> is the specified name prefix,
+ * - <version> is an unknown version that needs to be discovered, and
+ * - <segment> is a segment number. 
+ * Note: The number of segments is unknown and is controlled by the
+ *       `FinalBlockId` field in the first retrieved Data packet. So, the
+ *       very first Data packet MUST contain the `FinalBlockId` field.
+ *
+ * The following logic is implemented in PipelineFixed:
+ *
+ * 1. Express the first Interest to discover the version:
+ *
+ *    >> Interest: /<prefix>?MustBeFresh=true
+ *
+ * 2. Infer the latest version of the Data: <version> = Data.getName().get(-2)
+ *
+ * 3. Pipeline the Interets starting from segment 0:
+ *
+ *    >> Interest: /<prefix>/<version>/<segment=0>
+ *    >> Interest: /<prefix>/<version>/<segment=1>
+ *    ...
+ *    >> Interest: /<prefix>/<version>/<segment=this.windowSize-1>
+ * 
+ * We do not issue interest for segments that are already received.
+ * At any given time the number of on the fly Interests should be equal
+ * to this.windowSize. The next expected segment to fetch will be this.windowSize
+ *
+ * 4. Upon receiving a valid Data back we pipeline an Interest for the
+ *    next expected segment.
+ *
+ * We repeat step 6 until the FinalBlockId === Data.getName().get(-1).
+ *
+ * 5. Call the onComplete callback with a Blob that concatenates the content
+ *    from all segments.
+ *
+ * If an error occurs during the fetching process, the onError callback is called
+ * with a proper error code.  The following errors are possible:
+ *
+ * - `INTEREST_TIMEOUT`: if any of the Interests times out (probably after a number of retries)
+ * - `DATA_HAS_NO_SEGMENT`: if any of the retrieved Data packets don't have a segment
+ *   as the last component of the name (not counting the implicit digest)
+ * - `SEGMENT_VERIFICATION_FAILED`: if any retrieved segment fails
+ *   the KeyChain verifyData.
+ * - `IO_ERROR`: for I/O errors when sending an Interest.
+ * -`DUPLICATE_SEGMENT_RECEIVED`: if any duplicate segment is receveid during
```

Comment:
* wrong indent
* typo 'receveid'
* why would this ever happen?

Reply:
Due to network congestion (and consequently delay). Assume we issue an interest and because of a congested network do not receive back the data, so we issue another one and this time congestion is gone. We can find a scenario (regarding to `expiryTimer`) that we receive 2 data packets for the same segment.